src/devtools/client/debugger/src/reducers/source-actors.js

```diff
@@ -101,16 +109,53 @@ function updateBreakableLines(state, action) {
   return updateResources(state, [{ id: sourceId, breakableLines: value }]);
 }
 
-export function resourceAsSourceActor({ breakpointPositions, breakableLines, ...sourceActor }) {
-  return sourceActor;
+function updateBreakpointHitCounts(state, action) {
+  const value = asyncActionAsValue(action);
+  if (value.state === "pending") {
+    return state;
+  }
+
+  const hitCountsMap = value.value.hits.reduce(
+    (acc, { location, hits }) => {
+      if (!acc.minColumn[location.line]) {
+        acc.hits[location.line] = hits;
+        acc.minColumn[location.line] = location.column;
+      } else if (acc.minColumn[location.line] > location.column) {
+        acc.hits[location.line] = hits;
+        acc.minColumn[location.line] = location.column;
+      }
+      return acc;
+    },
+    { hits: {}, minColumn: {} }
+  );
+
+  const { id: sourceId } = action;
+
+  if (!hasResource(state, sourceId)) {
+    return state;
+  }
+
+  return updateResources(state, [{ id: sourceId, breakpointHitCounts: hitCountsMap.hits }]);
+}
+
+function clearBreakpointHitCounts(state) {
+  const withoutBreakpointHitCounts = makeReduceAllQuery(
```

Comment:
:hand: FWIW I personally find this particular function a bit confusing / hard to read.  (Then again, I also just searched the codebase for `makeReduceAllQuery` and discovered the set of "make cached resource" functions for the first time, so I had no clue those existed until 2 minutes ago :)  )

BTW, as a side note: now that we've got Redux Toolkit in the codebase, we've got Immer available to us, which we can use to simplify any immutable update logic.  I started to see what this snippet might look like with Immer and decided it isn't as helpful here because we're converting the array into a lookup table anyway, but FYI that it's available for other spots where we're doing immutable updates.

Reply:
haha, two things: *I* had never seen this function before last night, and I could not agree more. I know how map reduce works, but I still just like sat scratching my head and looking at it for ten minutes. But I wanted to follow the established patterns as much as I could, so I'm sticking with it until we can do an overhaul on this reducer. And *yes* to immer, I have not used it a ton, but where I have it's been great, excited to have that on this project.